api-test/hasura/actions/_handlers/createEpoch.test.ts

```diff
@@ -0,0 +1,553 @@
+            frequency_unit: 'months',
+          },
+        })
+      );
+      const { start_date, end_date } = epoch;
+      expect(
+        Interval.fromISO(start_date + '/' + end_date).length('months')
+      ).toBe(DURATION_IN_MONTHS);
+    });
+
+    test('cannot repeat with a frequency that is shorter than the duration', async () => {
+      expect.assertions(1);
+      const now = DateTime.now();
+      const thunk = async () =>
+        client.mutate({
+          createEpoch: [
+            {
+              payload: {
+                circle_id: circle.id,
+                params: {
+                  type: 'custom',
+                  start_date: now.toISO(),
+                  end_date: now.plus({ days: 3 }).toISO(),
+                  frequency: 1,
+                  frequency_unit: 'days',
+                },
+              },
+            },
+            {
+              __typename: true,
+              epoch: {
+                start_date: true,
+                end_date: true,
+                repeat_data: [{}, true],
+              },
+            },
+          ],
+        });
+      await expect(thunk).rejects.toThrow('is longer than chosen');
+    });
+
+    test('errors on malformed input', () => {
+      let result = zEpochInputParams.safeParse({
+        type: 'custom',
+        start_date: 'a',
+        end_date: 'b',
+        frequency: 'bad',
+        frequency_unit: 'years',
+      });
+      if (result.success === false)
+        expect(result.error.format()).toMatchObject({
+          start_date: {
+            _errors: [
+              'invalid datetime: the input "a" can\'t be parsed as ISO 8601',
+            ],
+          },
+          end_date: {
+            _errors: [
+              'invalid datetime: the input "b" can\'t be parsed as ISO 8601',
+            ],
+          },
+          frequency: { _errors: ['Expected number, received nan'] },
+          frequency_unit: {
+            _errors: [
+              'Invalid literal value, expected "days"',
+              'Invalid literal value, expected "weeks"',
+              'Invalid literal value, expected "months"',
+            ],
+          },
+        });
+      result = zEpochInputParams.safeParse({
+        type: 'custom',
+        bad: 'input',
+      });
+      if (result.success === false)
+        expect(result.error.format()).toEqual({
+          _errors: ["Unrecognized key(s) in object: 'bad'"],
+          start_date: { _errors: ['Required'] },
+          end_date: { _errors: ['Required'] },
+          frequency: { _errors: ['Expected number, received nan'] },
+          frequency_unit: {
+            _errors: [
+              'Invalid literal value, expected "days"',
+              'Invalid literal value, expected "weeks"',
+              'Invalid literal value, expected "months"',
+            ],
+          },
+        });
+    });
+  });
+
+  describe('monthly input', () => {
+    test.todo('errors on malformed input');
```

Comment:
@topocount should we remove these and add them with the monthly follow on PR?

Reply:
@crabsinger It's not necessary because they are part of jest's api and are simply logged to the console. Nothing wrong with leaving them in for now imo.